socrata/socrata_upload.py

```diff
@@ -72,11 +75,30 @@ def build_query(athena_asset, row_identifier, years=None, township=None):
     passed to `row_identifiers`.
     """
 
-    row_identifier = f"CONCAT({', '.join(row_identifier)}) AS row_id"
+    if len(row_identifier) > 1:
+        row_identifier = f"CONCAT(CAST({' AS varchar), CAST('.join(row_identifier)} AS varchar)) AS row_id"
+    else:
+        row_identifier = f"CAST({row_identifier[0]} AS varchar) AS row_id"
 
     # Retrieve column names and types from Athena
     columns = cursor.execute("show columns from " + athena_asset).as_pandas()
 
+    # Limit pull to columns present in open data asset
+    asset_url = (
+        "https://datacatalog.cookcountyil.gov/resource/"
+        + asset_id
+        + ".json?$limit=1"
+    )
+
+    asset_columns = (
+        requests.get(
+            asset_url, headers={"X-App-Token": os.getenv("SOCRATA_APP_TOKEN")}
+        )
+        .json()[0]
+        .keys()
+    )
+    columns = columns[columns["column"].isin(asset_columns)]
+
```

Comment:
Can't pass the API columns it isn't looking for.

Reply:
**[Thought, non-blocking]** This works I suppose, but it feels brittle that we need to send a web request to the asset to figure out what columns we want to push to it. It seems to me that our code should be the authority about what goes to the asset rather than the other way around. Plus, if this request fails, the whole script will fail; we could handle this by adding retry logic, but then we're potentially building in a bunch of complexity just so we can get information we should already know about the asset.

I wonder if there's a deeper, structural fix we could implement for restricting the columns that we send to open data? Some ideas:

* Add an optional list of columns to include under a key like `meta.columns` in the exposure configuration
* Create new views under a dedicated schema like `open_data` that build on the `default` views but restrict to only the data that we want to upload to open data

I'm happy to merge this in as-is to get things working, especially since it's not totally clear to us yet whether this is going to work at all, but in the long term I bet there's a more robust foundation we can build on.

Reply:
I've been thinking about this as well, and there's going to be another problem that pops up very similar to the one you're outlining - we have changed API field names for some assets, and the API is going to crap out because of that. Either we need to make sure every api field name is that same as the column name in the athena view that feeds the asset, or we need to create a crosswalk table similar to what you're outlining above.

My current thought it this - creating these athena tables is probably the more robust approach, and I've considered it, but I'm also deeply unsure that this API solution is actually going to end up being the solution we were hoping it might be given the issues I'm having with being kicked out by socrata's server due to time or memory. If I can resolve those issues and get to the point where I'm able to update an entire year of just the sales asset, it'll be worth sketching out a better solution to the column and api field name issue.